File: cs/Graphs/Network.cs
using System;
using System.Collections.Generic;

namespace Algorithms.Graphs
{
	public class Network
	{
		private readonly List<int>[] _edges;
		private readonly int _source, _sink;
		private readonly int[,] _capacity, _flow;
		private readonly int[] _tag;
		private int _cookie = 1, _totalFlow;
	
		public Network(int maxNodes)
		{
			_source = maxNodes;
			_sink = maxNodes + 1;
			maxNodes += 2;
			_edges = new List<int>[maxNodes];
			for(int i=0; i<maxNodes; i++) _edges[i] = new List<int>();
			_capacity = new int[maxNodes,maxNodes];
			_flow = new int[maxNodes,maxNodes];
			_tag = new int[maxNodes];
		}

		public void AddSourceEdge(int dest, int capacity) { AddEdge(_source, dest, capacity); }
		public void AddSinkEdge(int src, int capacity) { AddEdge(src, _sink, capacity); }

		public void AddEdge(int src, int dest, int capacity)
		{
			if (_capacity[src,dest] == 0 && _capacity[dest,src] == 0)
			{
				if (src != _sink) _edges[src].Add(dest);
				if (src != _source) _edges[dest].Add(src);
			}
			_capacity[src,dest] += capacity;
		}

		private int Go(int cur, int flow)
		{
			if (cur == _source) _cookie++;
			if (cur == _sink) return flow;
			if (_tag[cur] == _cookie || flow == 0) return 0;
			_tag[cur] = _cookie;
			foreach(int v in _edges[cur])
			{
				int f = Go(v,Math.Min(flow, _capacity[cur,v] - _flow[cur,v] + _flow[v,cur]));
				if (f > 0) { _flow[cur,v] += f; return f; }
			}
			return 0;
		}

		public int Flow()
		{
			int f;
			while ((f = Go(_source, int.MaxValue)) > 0) _totalFlow += f;
			return _totalFlow;
		}
	}
}

File: java/src/yarin/yal/graph/Network.java
package yarin.yal.graph;

import java.util.ArrayList;
import java.util.List;

public class Network {
    private final List<List<Integer>> edges;
    private final int source, sink;
    private final int[][] capacity, flow;
    private final int[] tag;

    private int cookie = 1, totalFlow;

    /**
     * Create new network
     * @param n number of nodes, excluding source and sink
     */
    public Network(int maxNodes)
    {
        source = maxNodes;
        sink = maxNodes + 1;
        maxNodes += 2;
        edges = new ArrayList<List<Integer>>(maxNodes);
        for(int i=0; i<maxNodes; i++) {
            edges.add(new ArrayList<Integer>());
        }
        capacity = new int[maxNodes][maxNodes];
        flow = new int[maxNodes][maxNodes];
        tag = new int[maxNodes];
    }

    public void addSourceEdge(int dest, int capacity) {
        addEdge(source, dest, capacity);
    }

    public void addSinkEdge(int src, int capacity) {
        addEdge(src, sink, capacity);
    }

    public void addEdge(int src, int dest, int capacity) {
        if (this.capacity[src][dest] == 0 && this.capacity[dest][src] == 0) {
            if (src != sink) {
                edges.get(src).add(dest);
            }
            if (src != source) {
                edges.get(dest).add(src);
            }
        }
        this.capacity[src][dest] += capacity;
    }

    private int go(int cur, int flow) {
        if (cur == source) {
            cookie++;
        }
        if (cur == sink) {
            return flow;
        }
        if (tag[cur] == cookie || flow == 0) {
            return 0;
        }
        tag[cur] = cookie;
        for (int v : edges.get(cur)) {
            int f = go(v, Math.min(flow, capacity[cur][v] - this.flow[cur][v] + this.flow[v][cur]));
            if (f > 0) {
                this.flow[cur][v] += f;
                return f;
            }
        }
        return 0;
    }

    public int flow() {
        int f;
        while ((f = go(source, Integer.MAX_VALUE)) > 0) {
            totalFlow += f;
        }
        return totalFlow;
    }
}
